### PR TITLE
Cleanup _plot_args_replacer logic

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

This is a follow-up to #10810. Basically it cleans up the code structure. The logic in the PR is the same, just rewritten to be more clear.

### Additional follow up question

As part of the cleanup, I realized that we are not exactly checking what we want to know:
- *Want to know* whether *data* as an item named *name*.
- *Check* `name in data or name in data.dtype.names`.

That works for usual data (dict like and numpy.array), but may fail e.g. if a user would use a string-list for data. I assume that's ok-ish.

I think there is no real "has_item()" function and the only exact way of finding out if `data[name]` works is really trying and catching a possible exception. I assume we don't want to really pull data as part of this check.

So in total the present check is still the check we want to do. Please correct me if I'm wrong here.

### CODE REVIEW

**Comment by phobson** [comment]
> may fail e.g. if a user would use a string-list for data. I assume that's ok-ish.

I'm worried this might mess categorical axes. I'll make a note to check on that.

**Comment by tacaswell** [comment]
This will 'work' if you do `ax.plot('foo', 'bar', data=['foo', 'bar'])`

**Comment by anntzer** [comment]
Why do we not want to pull out the data as part of the test? (i.e. try `data[name]` and catch attributeerror) "Should" be cheap...

**Comment by timhoffm** [comment]
"Probably" you are right.

In the worst case, it could get a factor 2 slower because we'll have to fetch the data later anyway. But I think this is negligible compared to all the other stuff done during plotting. If someone has data with a slow getitem, they can still resort to `plot(data['x'], data['y'])` to only access them once.

So I'm +1 on try `data[name]` and catch attributeerror.

**Comment by timhoffm** [comment]
Is there more support for test by access? If so, I would create an alternative PR.

**Comment by tacaswell** [comment]
h5py objects can be slow.

I would lean towards continuing to use `__contains__` and documenting that support that is part of the required API on the things passed in to `data=`.

**Comment by timhoffm** [comment]
If h5py is the couter-example that item access can be slow, I'd leave the implementation logic as is, just with the cleaned up code from this PR.

I've added a note that the data object must support membership test.

File: lib/matplotlib/__init__.py [hunk 1]
```diff
@@ -1529,6 +1529,9 @@ def _replacer(data, key):
     following arguments are replaced by **data[<arg>]**:
 
     {replaced}
+    
+    Objects passed as **data** must support item access (``data[<arg>]``) and 
+    membership test (``<arg> in data``).
```
Comment (efiring) [inline lib/matplotlib/__init__.py]: This is inconsistent with the actual test, and with the desired behavior; the structured ndarray does not support the membership test.

Comment (tacaswell) [reply lib/matplotlib/__init__.py]: Maybe should say 'support `<arg> in data` or be a numpy structured array`?

**Comment by anntzer** [comment]
I don't mind the rebase, but note that #10928 will probably make all this obsolete.

**Comment by tacaswell** [comment]
Don't we need both?

**Comment by anntzer** [comment]
#10928 kills _plot_args_replacer and uses the standard _replacer logic used by everyone else (try to get the item and if it's there, swap it in -- so access is only done one anyways)

**Comment by timhoffm** [comment]
Note: Adapted `_has_item()` to fix #11389

**Pull request merged**